docs/architecture/adrs/002-streaming-response-aggregation.md

```diff
@@ -0,0 +1,43 @@
+# ADR 002: Streaming Response Behavior and Result Aggregation
+
+This ADR defines the orchestrator strategy of result aggregation and behavior observed by end users of the initial streaming API endpoint (i.e. streaming text generation with classification) when detections are made on streamed output of text generation models.
+
+## Motivation
+
+This orchestrator is designed to work with various detectors. Each detector will have an associated chunker indicated in an end user configuration. Each chunker could chunk different amounts of text e.g. a sentence chunker vs. a paragraph chunker. Each detector could then provide detection responses on different amounts of text. In the unary case, all of these detection responses could be aggregated together and returned in the single response to the end user. 
+
+However, in the streaming case, as frames are being streamed back, the end user will need to receive detection responses corresponding to the text in the streamed frames. Frames still have to be returned in order so that the user can parse the stream of generated text and detections and know how much of the original text generation stream has been processed i.e. what part of the stream that detections have been run on.
+
+
+## Decisions
+
+### End user experience
+
+1. Only fully processed frames will be returned to the end user. This means that for each (streamed) frame returned to the user, all detector responses for that particular frame's text up to that frame’s index (`processed_index`) will be returned with the frame.
+    1. Generated texts in output frames with detections are still expected to be contiguous, meaning the end user can still piece all of the frames together. This aligns with the way various text generation servers provide output.
+2. In order to achieve (1) for detectors that may work with different chunkers, e.g. paragraph vs. sentence, results will not be streamed back/returned until the “slowest” detector has processed. “Slowest” may not necessarily just be in terms of time but in terms of amount of text processed (i.e. largest `processed_index`).  
+
+### Response aggregation strategy
+
+1. Only chunker calls will provide indices of the text generation server frames processed (`start_index` and `processed_index`). Detector calls are expected to be unary, and responses only indicate the detections on any given chunk from a chunker. Detectors are expected to still give a reply (i.e. empty array) if there are no detections on a given chunk.
+    1. On any detector response, the orchestrator will have to associate the chunker indices with each detection response. The orchestrator will update any span start/end on detection responses with chunker indices to be able to indicate to the end user where the detection was in the overall stream.
+2. The orchestrator will have to maintain detection responses that are not ready to be streamed back to the user, such as with queues.
```

Comment:
Can we make a note here about how this will work along with thresholding? Will we store results after applying threshold in the queue or before applying threshold?

Reply:
Yeah, I think thresholding them first make sense.  But in that case, would we not put it in queue if the response turns out to be not relevant ? Since in that case, how would be know that the detector has responded, because then there are 2 scenarios that can cause the response to be not in the queue, i.e detector has not responded and detector responded but the response was not relevant. In both cases, we would end up waiting ?

Reply:
sounds good! Thanks